In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
visa_df=pd.read_csv(r'C:\Users\MOHD RAHEEM\Documents\Naresh IT\Data-Files\Visadataset.csv')
visa_df.head(2)

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified

#### Standardization:
- Standardization is a technique that scales all the data under one scale
- Different columns has different values also different units
- Some columns values has bigger values, some column values has lesser values
- So it is important to keep all the values under one scale
- We have two methods are there
    - Standardization
    - Normalization
- Standardization
    - Z-score
    - Z-score values varies from -3 to 3

<img alt="" style="width: 301.9px; height: 169.8px; margin-top: 5.1px; margin-left: 0px;" class="" src="https://tse2.mm.bing.net/th?id=OIP.1TFd1ElbBIGaIoxAGgWDigHaEK&amp;pid=Api&amp;P=0&amp;h=180" id="yui_3_5_1_1_1718116491618_681">

#### Normalization:
- min max scalar
- values are ranges from 0 to 1
<img alt="" style="width: 699.1px; height: 165.9px; margin-top: 7.1px; margin-left: 0px;" class="" src="https://tse1.mm.bing.net/th?id=OIP.GmFes4H0-rk6zdI5YQAM1wHaBw&amp;pid=Api&amp;P=0&amp;h=180" id="yui_3_5_1_1_1718116679586_681">

#### Task:
- Step1: Take prevailing wage column: **visa_df['prevailing_wage']**
- Step2: Calculate the mean value of prevailing wage: mean= **visa_df['prevailing_wage'].mean()**
- Step3: Calculate the standard deviation of prevailing wage: std= **visa_df['prevailing_wage'].std()**
- Step4: Calculate the Nr: step-1 - step2: **Nr=visa_df['prevailing_wage']-mean**
- Step5: Divide the **step4/step3**

In [4]:
dict1={'Name':'Amer','Age':[19,10]}
pd.DataFrame(dict1)

Name  Age
0  Amer   19
1  Amer   10

In [6]:
mean=visa_df['prevailing_wage'].mean()
std=visa_df['prevailing_wage'].std()
Nr=visa_df['prevailing_wage'] - mean
visa_df['prevailing_wage_z']=Nr/std
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  prevailing_wagae  prevailing_wage_z  
0          Denied         -1.398510          -1.398510  
1       Certified          0.169832           0.169832  
2          Denied          0.919060           0.919060  
3          Denied          0.169991           0.169991  
4       Certified          1.428576           1.428576  
...           ...               ...                ...  
25475   Certified          0.049923           0.049923  
25476   Certified          3.876083           3.876083  
25477   Certified          1.360253           1.360253  
25478   Certified          0.221504           0.221504  
25479   Certified         -0.067762          -0.067762  

[25480 rows x 14 columns]

#### Task-2:
- Compare the those two columns i.e, **visa_df['prevailing_wage']**  &  **visa_df['prevailing_wage_z']**  
- Get the indexes of both maximum value and minimum value of the column,
- The both indexes should match.

#### .idxmax() & .idxmin()
- Since **.max()** & **.min()** are used to calculate the **maximum** & **minimum** value of particular column.
- But in order to get the indexes(idx) of those **max_val** & **min_val** we use **.idxmax()** & **.idxmin()** respectively.

In [7]:
maxx=visa_df["prevailing_wage"].idxmax(),visa_df["prevailing_wage_z"].idxmax()
minn=visa_df["prevailing_wage"].idxmin(),visa_df["prevailing_wage_z"].idxmin()

maxx,minn

((21077, 21077), (20575, 20575))

- So **21077** is the index position for maximum value of the column,
- **20575** is the index position for minimum value of the column.

In [8]:
max_val=visa_df["prevailing_wage"].max(),visa_df["prevailing_wage_z"].max()
min_val=visa_df["prevailing_wage"].min(),visa_df["prevailing_wage_z"].min()

max_val,min_val

((319210.27, 4.634101837909902), (2.1367, -1.4096818992891214))

- This means 319210.27 **(max_Value)** is at 21077 **(index_position)**
- This means 2.1367 **(max_Value)** is at 20575 **(index_position)**

#### StandardScalar:
- StandardScalar same as Z-score but by using package
- It is under sklearn package
- In the sklearn we have preprocessing
- Read the data again 
- Read the package
- Save the package
- Apply fit transform 
- Compare 3 columns
    - One is original **visa_df['prevailing_wage']** 
    - Manually we did z-score **visa_df['prevailing_wage_z']**
    - Column with package **visa_df['prevailing_wage_ss']**

In [9]:
from sklearn.preprocessing import StandardScaler
ss=StandardScaler()
visa_df['prevailing_wage_ss']=ss.fit_transform(visa_df[['prevailing_wage']])

- Single square bracket is series
- Double square bracket is Dataframe
- Whenever you see the shape error apply double square bracket

In [10]:
visa_df[['prevailing_wage','prevailing_wage_z','prevailing_wage_ss']]

prevailing_wage  prevailing_wage_z  prevailing_wage_ss
0             592.2029          -1.398510           -1.398537
1           83425.6500           0.169832            0.169835
2          122996.8600           0.919060            0.919079
3           83434.0300           0.169991            0.169994
4          149907.3900           1.428576            1.428604
...                ...                ...                 ...
25475       77092.5700           0.049923            0.049924
25476      279174.7900           3.876083            3.876159
25477      146298.8500           1.360253            1.360280
25478       86154.7700           0.221504            0.221509
25479       70876.9100          -0.067762           -0.067763

[25480 rows x 3 columns]

#### Now,
### (2) Normalization:
#### Task (Do manually):
- Read the data again
- Step1: Take prevailing wage column:**visa_df['prevailing_wage']**
- Step2: Calculate the min value prevailing wage: min=**visa_df['prevailing_wage'].min()**
- Step3: Calculate the max value of prevailing wage: max=**visa_df['prevailing_wage'].max()**
- Step4: Calculate the Nr: step1 - step2: Nr=**visa_df['prevailing_wage']-min**
- Step5: DR=step3-step2
- Step6: Divide the step4/step5

In [11]:
x_max=visa_df['prevailing_wage'].max()
x_min=visa_df['prevailing_wage'].min()
Nr=visa_df['prevailing_wage'] - x_min
visa_df['prevailing_wage_min_max']=Nr/(x_max-x_min)
visa_df[['prevailing_wage','prevailing_wage_min_max']]

prevailing_wage  prevailing_wage_min_max
0             592.2029                 0.001849
1           83425.6500                 0.261345
2          122996.8600                 0.385312
3           83434.0300                 0.261371
4          149907.3900                 0.469616
...                ...                      ...
25475       77092.5700                 0.241505
25476      279174.7900                 0.874579
25477      146298.8500                 0.458311
25478       86154.7700                 0.269895
25479       70876.9100                 0.222033

[25480 rows x 2 columns]

#### Task (Use package & do):
#### package name: MinMaxScalar

In [12]:
from sklearn.preprocessing import MinMaxScaler
mms=MinMaxScaler()
visa_df['prevailing_wage_min_max_ss']= mms.fit_transform(visa_df[['prevailing_wage']])

visa_df[['prevailing_wage','prevailing_wage_min_max','prevailing_wage_min_max_ss']]

prevailing_wage  prevailing_wage_min_max  prevailing_wage_min_max_ss
0             592.2029                 0.001849                    0.001849
1           83425.6500                 0.261345                    0.261345
2          122996.8600                 0.385312                    0.385312
3           83434.0300                 0.261371                    0.261371
4          149907.3900                 0.469616                    0.469616
...                ...                      ...                         ...
25475       77092.5700                 0.241505                    0.241505
25476      279174.7900                 0.874579                    0.874579
25477      146298.8500                 0.458311                    0.458311
25478       86154.7700                 0.269895                    0.269895
25479       70876.9100                 0.222033                    0.222033

[25480 rows x 3 columns]